Fix Protractor tests in Travis

kaidohallik opened the pull request:
Fixes #10806

Seems that something changed in the latest `webdriver-manager` version `12.1.7` to what our clients were updated by #10676. Seems that if UI element is not inside visible area then `Protractor` is not able to get reference to that element any more.

In this PR I changed Chrome window 10 times higher and now Protractor tests are passing.

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

<!--
Please also reference the issue number in a commit message to [automatically close the related Github issue](https://help.github.com/articles/closing-issues-via-commit-messages/)

Note: It is also possible to add `[skip ci]` to your commit message to skip Travis tests
-->

kaidohallik commented:
Hmm, https://travis-ci.org/jhipster/generator-jhipster/jobs/601939724 passed in #10676 - maybe then Chrome latest version introduced this behavior change at some point: https://github.com/jhipster/generator-jhipster/blob/master/.travis.yml#L32

And locally can these people reproduce, who are using latest Chrome?

Problem starts if entities menu is high enough so that some menu items are not visible after opening entities menu.

MathieuAA commented:
Awesome!!

pascalgrimaud commented:
@kaidohallik : wow ! are you confident ? Can I close / reopen to trigger the build plz ?

kaidohallik commented:
I see that builds passed, I re-runned initially failed GitHub tests at some point (error was that directory existed in /tmp folder, so temporary glitch), maybe you are seeing cached view?

pascalgrimaud commented:
So only Angular needs fix, not React...
Then, let's merge this, it will be better than what we currently have !

kaidohallik commented:
Is some React build using sql full profile?  
It's possible that this is related only to Angular.  
One more thought: failed only that's entity test which menu item was under scrollbar.  
Menu items before (visible) and after scrollbar (hidden) were successfully caught by Protractor.

pascalgrimaud commented:
@kaidohallik : no, we only have sqlfull for angular, not for react
- ngx : https://github.com/hipster-labs/jhipster-daily-builds/blob/master/azure-pipelines-ngx-maven.yml#L54
- react : https://github.com/hipster-labs/jhipster-daily-builds/blob/master/azure-pipelines-react-maven.yml#L53

I'm going to change it

sendilkumarn commented:
@kaidohallik Interesting find and fix 👏 

@pascalgrimaud can we do a minor release 😬 ? I will release KHipster on top of that :)

pascalgrimaud commented:
@sendilkumarn : I can, but I'm not sure it is needed as the issue only affects our CI.

I didn't manage to reproduce it locally. @kaidohallik did you ?

Anyway, I'm not sure it's the right fix, changing 1024 to 10240 looks a bit strange... but as it makes the CI passed, it's ok for me, for now.

The real fix should be to upgrade the webdriver maybe ?

kaidohallik commented:
@pascalgrimaud Yes, I reproduced it locally, My Chrome version in Windows 10 is 78.0.3904.108

According to tests in my machine and in Travis it affects users who has 31 or more entities, without this fix their Protractor tests will fail with current Chrome latest stable.

We can try to find lower value instead of 10240, but as this window size must be high enough so that any normal monitor can't show the whole Chrome window then it doesn't matter is this a bit lower or higher.

We are using latest webdriver-manager, version 2.1.7. I tried downgrading to 2.1.6 but it's not possible because only latest 2.1.7 supports latest Chrome version 78.

Maybe some future version of Chrome or webdriver-manager fixes this problem and we can revert this change.

I tried to use Chrome Canary (version 80.0.3974.0) but ChromeDriver (version 78.0.3904.105) complained and not success with that: This version of ChromeDriver only supports Chrome version 78

And finally one correction to my previous post: I said that menu item under scrollbar is not accessible to Protractor but in fact this specific menu item not accessible is slightly below the border of the Chrome window.

vishal423 commented:
Instead of providing a fixed window size, can we occupy complete screen size with `browser.driver.manage().window().maximize();`

kaidohallik commented:
With `browser.driver.manage().window().maximize()`;
* 2560 x 1440 monitor - success
* 1920 x 1080 monitor - failed

I made several other tests with different browser window size and different menu item ordering and below is the summary.

Noticed in the manual testing:
* entities menu is opened at the right border of the browser window (i18n is disabled so no Languages menu which helps to push menu a little bit to left)
* moving mouse over those entities menu items whose label is near to maximum entity label length causes horizontal scrollbar to appear in the browser

So maybe:
* if menu item is out of the visible area
* and on mouse over this menu item causes horizontal scrollbar to appear
* then Protractor is not able to click on that menu item

vishal423 commented:
> 1920 x 1080 monitor - failed

I think we should support this resolution.

> then Protractor is not able to click on that menu item

If it works in normal testing, then, we may not be triggering a click event on correct element.

kaidohallik commented:
Maybe in e2e tests the following scenario:
* Protractor using Chrome calculates coordinates where to click
* and if doing click, first on mouse over action triggers and scrollbar appears
* this changes wanted menu item y coordinate by scrollbar height
* and if scrollbar is big enough then click is performed out of the clickable area

kaidohallik commented:
I made the following test:
* I restored `Pager` entities by reverting #10760, so that generated `navbar.html` would be identical to what it was before #10760 (and removed failing relationships from `TestPager`)
* I reverted window height change, 10240 --> 1024

And Travis tests successful:
* branch: https://github.com/kaidohallik/generator-jhipster/tree/travis-with-pager
* successful Travis test: https://travis-ci.org/kaidohallik/generator-jhipster/builds/615520051

So conclusion from here is that this `EntityWithServiceClassPaginationAndDTO` e2e test failure was not caused by new Chrome version as I thought.

Seems that instead this is related to how exactly entities menu looks like (and what are browser parameters).

If we want to avoid this browser height 10240 and want to revert that back to 1024 then probably one possible solution is to increase entities menu right padding or something similar to avoid scrollbar on mouse over in long labeled menu items.

kaidohallik commented:
https://github.com/jhipster/generator-jhipster/blob/master/generators/client/templates/angular/src/main/webapp/content/scss/global.scss.ejs#L38 - on mouse over the menu item text is changing to bold and the longer the text is the more free space is needed in the right side of the menu to avoid scrollbar.

kaidohallik commented:
I made test:
* reverted window height change, 10240 --> 1024
* removed `bold` from `a:hover` in `global.scss`

And Travis passed: https://travis-ci.org/kaidohallik/generator-jhipster/builds/615610529

So this seems to confirm theory that scrollbar that appears on long menu item hover (which is caused by changing font to bold) is causing Protractor failures in certain situations.

Seems that bold is used only in default theme to resolve this Bootstrap issue with default theme: https://getbootstrap.com/docs/4.1/getting-started/accessibility/#color-contrast and to get 100% on Lighthouse Accessability Audit.

So `e2e` is probably failing only in default scheme.

Lately #10102 dealt with this problem, but `bold` was not removed from `a:hover`.
